system/ehci.c

```diff
@@ -437,6 +437,24 @@ static bool get_data_request(const usb_hcd_t *hcd, const usb_ep_t *ep, const usb
     return do_async_transfer(ws, 3);
 }
 
+// FIXME
+static bool out_data_request(const usb_hcd_t *hcd, const usb_ep_t *ep, const usb_setup_pkt_t *setup_pkt,
+                             const void *buffer, size_t length)
+{
+    workspace_t *ws = (workspace_t *)hcd->ws;
+
+    if (setup_pkt) {
+        build_ehci_qtd(&ws->qtd[0], &ws->qtd[2], EHCI_QTD_PID_SETUP, EHCI_QTD_DT(0), setup_pkt, sizeof(usb_setup_pkt_t));
+        build_ehci_qtd(&ws->qtd[1], &ws->qtd[2], EHCI_QTD_PID_IN,    EHCI_QTD_DT(1), NULL, 0);
+        build_ehci_qtd(&ws->qtd[2], &ws->qtd[2], EHCI_QTD_PID_OUT,   EHCI_QTD_DT(1), buffer, length);
+    } else {
+        // what DT ?
+        build_ehci_qtd(&ws->qtd[0], &ws->qtd[0], EHCI_QTD_PID_OUT,   EHCI_QTD_DT(1), buffer, length);
+    }
+    build_ehci_qhd(&ws->qhd[0], &ws->qtd[0], ep, false);
+    return do_async_transfer(ws, 3);
```

Comment:
`do_async_transfer(ws, 3)` even after a single call to `build_ehci_qtd()` in the `else` path raised my eyebrows. Now, I'm not knowledgeable about USB controller usage, and this can be entirely justified, but in that case, I'd say that this warrants a comment :)

Reply:
The EHCI controller provides separate schedulers for asynchronous (control and bulk) and periodic (isochronous and interrupt) transfers. So this is something inherent in how the EHCI controller operates.

Reply:
Note the ehci changes, as per the commit message, hasn't been tested. For instance I am not sure what the DT value is meant to be, also as commented in the code. The point is that for bulk transfers there is only one packet going out (with ACK from device), whereas for control transfers there is the setup and in/out stages. The rest is copied (cargo-culted) from the existing get_data_request().